python/aitemplate/compiler/transform/fuse_split.py

```diff
@@ -243,11 +247,25 @@ def _fuse_split_and_strided_op(sorted_graph: List[Tensor]) -> List[Tensor]:
         # still have mis-aligned accesses caused by offsets. This _check_alignment
         # filters out all bad cases.
         for output in outputs:
+            total_elems_from_split_dim = (
+                stride * split_input._attrs["shape"][split_dim].value()
+            )
             can_fuse_split &= len(output.dst_ops()) > 0 and all(
-                _is_supported_op(next_op._attrs["op"])
-                # need to pass the real offset to alignment checker
-                and _check_alignment(next_op, dim_offset * stride)
-                and len(output.dst_ops()) == 1
+                (
+                    _is_supported_op(next_op._attrs["op"])
+                    # need to pass the real offset to alignment checker
+                    and _check_alignment(
+                        next_op, dim_offset * stride, total_elems_from_split_dim
+                    )
+                    and len(output.dst_ops()) == 1
+                )
+                or (
+                    next_op._attrs["op"] == "concatenate"
+                    and next_op._attrs["concat_dim"] == split_dim
+                    and _check_alignment(
+                        next_op, dim_offset * stride, total_elems_from_split_dim
+                    )
```

Comment:
I think your example of output used by both concat and elementwise would already be handled, since the check requires `all` of `output.dst_ops` is a supported bmm (which enforces 1 dst_op only) or concat (which doesn't check 1 dst_op only). If `split_2[1]` is used by both concat and add, then it will fail the `next_op._attrs["op"] == "concatenate"` check and the optimization will not be done.

Actually, I am not sure why the `len(output.dst_ops()) == 1` check is needed in the bmm case either, but I do remember a test failing when I tried having the same split output used by both bmm and concat.

Reply:
I think you are right. Thanks for the explanation!